### DISCUSSION TITLE
control characters in Trino UDF

### DISCUSSION THREAD

**Discussion by natarajank11** [trinodb/trino#17676] [General]
Trino 407, I am writing UDF function to remove all the control characters from the given string.

`

    public static Slice rmcc(@SqlType(StandardTypes.VARCHAR) Slice string)
    {
        String text = string.toStringUtf8();

        String regex = "[\\\\p{C}]";
        Pattern p = Pattern.compile(regex);
        Matcher m = p.matcher(text);
        String result = m.replaceAll("");

        return Slices.utf8Slice(result);
     }

`

However the control characters are not removed in Trino CLI,
`select rmcc('\b\t\f\0'); => btf0`

**Comment by hashhar** [comment]
I'd suggest you to check how the literal looks like before writing the regex.

`SELECT '\b\t\f\0'` will show you that each of those "control" characters are treated literally - i.e. they have no special meaning in SQL.

So you cannot use the control character regex to match them since they are not control characters - it's literally a string `'\' + 'b' + '\' + 't' + '\' + '0'`. So you regex needs to be `String regex = "\\[\\b\\t\\f\\0]"`. Hopefully you see the problem now - the string doesn't have any control characters so removing them makes no sense.

For example does this have a control character or it's a valid path? `C:\Users\bob\tom\foo\000\file`?

**Reply by natarajank11** [reply]
understand, I would like to know how to handle all the control characters in regex instead of mentioning particular characters.

**Reply by hashhar** [reply]
My point is there is no control character in varchar literals so you cannot say "handle all control characters". You need to define what characters you want to replace. an example is to replace all occurrences of `\` + any alphabet which can be represented as the regex `String regex = "\\\\[a-zA-Z]"`.
